src/timely-util/src/operators_async_ext.rs

```diff
@@ -182,7 +189,10 @@ macro_rules! async_op {
                 let mut $capabilities = &mut capabilities;
                 #[allow(unused_mut)]
                 let mut $frontiers = &mut frontiers;
-                let _: () = async { $body }.await;
+
+                if !async { $body }.await && frontiers.borrow().iter().all(|f| f.is_empty()) {
```

Comment:
Since this macro is managing the termination condition on its own why do we need the body to signal with a boolean?

I think the expectation of the code fragment passed to the `async_op!` is that it will always process the inputs whenever it is running and `async_op!` will know by itself when to stop. So concretely I was expecting something like this here:

```rust
async { $body }.await;

if frontiers.borrow().iter().all(|f| f.is_empty())  {
    break;
}

```